onnxruntime/core/session/inference_session.cc

```diff
@@ -810,6 +810,20 @@ std::pair<common::Status, const InputDefList*> InferenceSession::GetModelInputs(
   return std::make_pair(common::Status::OK(), &model_->MainGraph().GetInputs());
 }
 
+std::pair<common::Status, const InputDefList*> InferenceSession::GetModelInputsIncludingInitializers() const {
+  {
+    std::lock_guard<onnxruntime::OrtMutex> l(session_mutex_);
+    if (!is_model_loaded_) {
+      LOGS(*session_logger_, ERROR) << "Model was not loaded";
+      return std::make_pair(common::Status(common::ONNXRUNTIME, common::FAIL, "Model was not loaded."),
+                            nullptr);
+    }
+  }
+
+  // return required inputs including any inputs used for overriding initializers
+  return std::make_pair(common::Status::OK(), &model_->MainGraph().GetInputsIncludingInitializers());
+}
```

Comment:
If the IR version is < 4 this will include values which are not allowed to be overridden. Prior to IR v4 all inputs had to have a matching initializers despite the intention for them to be treated as constant. We now enforce this and don't allow overriding an initializer with a matching graph input if the IR version < 4.

Not sure if we want to filter out such values. If so, should call Graph::CanOverrideInitializer() and just return the list from GetInputs if that is false.

Reply:
We should discuss and agree on the approach as this will impact what a lot of other function names/documentation should have. 

---
In reply to: [324997257](https://github.com/microsoft/onnxruntime/pull/1854#discussion_r324997257) [](ancestors = 324997257)

Reply:
Yes, I was thinking about it and there is also a input_defs map that actually calls CanOverrideInitializer() prior to being populated. One approach would be not to introduce new APIs at all but return all inputs for IR 4 and no initializers prior to IR4. That way the changes would be minimal and no changes required for C/C++ API or python.

---
In reply to: [325016736](https://github.com/microsoft/onnxruntime/pull/1854#discussion_r325016736) [](ancestors = 325016736,324997257)

Reply:
Hi, @yuslepukhin ,  For IR version > 4, GetInputsIncludingInitializers() returns (inputs - initializers).  So it's not what you want. Based on the offline talk, I guess you want the inputs and all the overridable initializers, indeed, the set is just the input list in the model file, without any filtering and extending.

Reply:
Please see graph.cc:2417

```C++
              if (!is_initializer || ir_version_ < 4) {
                graph_inputs_including_initializers_.push_back(input_arg);
              }
```
If the ir_version >= 4, it become:
```
if (!is_initializer) {
            graph_inputs_including_initializers_.push_back(input_arg);
}
```

Reply:
> Hi, @yuslepukhin , For IR version > 4, GetInputsIncludingInitializers() returns (inputs - initializers). So it's not what you want. Based on the offline talk, I guess you want the inputs and all the overridable initializers, indeed, the set is just the input list in the model file, without any filtering and extending.

I think what he has is right. For IR version >=4,  GetInputsIncludingInitializers() returns inputs + initializers that can be overridden.

Reply:
> We should discuss and agree on the approach as this will impact what a lot of other function names/documentation should have.
> 
> In reply to: [324997257](https://github.com/microsoft/onnxruntime/pull/1854#discussion_r324997257) [](ancestors = 324997257)

yeah, seems all kind of confusing.
from user perspective, they just want to know what the required inputs are and what the optional inputs are. there shouldn't be overlap between what the api's return. (so i'm not a fan of GetInputsIncludingInitializers() )

Reply:
I have tested what I have with Python and it seems right. With regards to the actual API, I am in favor of what Scott proposed. Preserving a single API GetInputs() and return initializers for IR >= 4.

---
In reply to: [325372721](https://github.com/microsoft/onnxruntime/pull/1854#discussion_r325372721) [](ancestors = 325372721)

Reply:
George and Pranav chatted and the proposal is to have two separate API 1) GetInputs() 2) GetInitializers()

---
In reply to: [325379931](https://github.com/microsoft/onnxruntime/pull/1854#discussion_r325379931) [](ancestors = 325379931,325372721)

Reply:
GetInputsIncludingInitializers is the complete list of values from GraphProto.inputs(), including graph inputs with backing initializers. 

The code you reference is if you are building a graph manually where SetGraphInputs may have been called. The 'normal' path involves 

    // Set graph inputs.
    // <graph_inputs_including_initializers_> contains inputs exactly specified in proto.
    // <graph_inputs_excluding_initializers_> contains inputs without default value (specified as initializer).
    for (auto& graph_input : graph_proto_->input()) {
      auto& name = graph_input.name();
      const auto* node_arg = GetNodeArg(name);
      ORT_ENFORCE(node_arg, "Graph ctor should have created NodeArg for initializer.");
      graph_inputs.insert({name, node_arg});
      graph_inputs_including_initializers_.push_back(node_arg);
      if (graph_initializers.end() == graph_initializers.find(name)) {
        graph_inputs_excluding_initializers_.push_back(node_arg);
      }
    }


---
In reply to: [325319403](https://github.com/microsoft/onnxruntime/pull/1854#discussion_r325319403) [](ancestors = 325319403)

Reply:
GetInputsIncludingInitializers lack of clarity is due to the lack of clarity in the ONNX spec regarding the treatment of initializers with matching graph inputs for IR version < 4. It made sense at the time, but the revision to treat initializers with matching graph inputs as constant if IR version < 4 changed that.

The proposed addition of GetOverridableInitializers would I think bring clarity. GetInputs returns required inputs. GetOverridableInitializers returns additional inputs for initializers that can be overridden. 


---
In reply to: [325372721](https://github.com/microsoft/onnxruntime/pull/1854#discussion_r325372721) [](ancestors = 325372721)